src/databroker_elasticsearch/elasticcallback.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
+
+"""\
+Callback for adding data to elastic search from run engine.
+"""
+
+from bluesky.callbacks.core import CallbackBase
+from databroker_elasticsearch.elasticindex import ElasticIndex
+
+
+class ElasticCallback(CallbackBase):
+    """
+    Callback for inserting start metadata into an `Elasticsearch` instance.
+
+    This wraps `ElasticIndex` adapter for adding translated documents to
+    a given Elasticsearch index.
+
+    Attributes
+    ----------
+    esindex : ElasticIndex
+        The ElasticIndex object for converting documents and inserting
+        them to Elasticsearch.
+    """
+
+
+    def __init__(self, esindex: ElasticIndex):
+        self.esindex = esindex
+        return
+
+
+    @classmethod
+    def from_config(cls, config):
+        """
+        Create a new ElasticCallback instance using a configuration dictionary.
+
+        Parameters
+        ----------
+        config : dict
+            The configuration dictionary that describes ElasticIndex.
+            It must contain "databroker-elasticsearch" key.
+
+        Returns
+        -------
+        ElasticCallback
+        """
+        from databroker_elasticsearch.elasticindex import ElasticIndex
+        esindex = ElasticIndex.from_config(config)
+        rv = cls(esindex)
+        return rv
+
+
+    def rebuild(self, headers, purge=False):
+        """Re-export start documents in given headers to Elasticsearch index.
+
+        Parameters
+        ----------
+        headers : iterable
+            The sequence of objects with "start" attribute which is exported.
+            This is usually an iterable of `databroker.Header` objects.
+        purge : bool, optional
+            When True purge the Elasticsearch index before adding headers.
+
+        Returns
+        -------
+        int
+            The number of documents that were added to Elasticsearch.
+        """
+        startdocs = (hdr.start for hdr in headers)
+        if purge:
+            self.esindex.reset()
+        cnt = self.esindex.devour(startdocs)
+        return cnt
+
+    # override CallbackBase function
+
+    def start(self, doc):
+        """Handle "start" document in bluesky run engine callback.
+        """
+        # FIXME should we cache and ingest in stop() to avoid canceled runs?
```

Comment:
I would ingest the canceled runs. If users want to filter by if a run has failed we can then maybe provide that functionality (maybe by also storing the stop documents?). We can always filter the results downstream, but we can only filter the things we have in the index.

Reply:
Sounds good.  We should add `status` or `complete` field to the ES document.
I will update the comment.

Reply:
I don't know if you would know that at start document time.
I would solve this usecase on the search side.

Reply:
We can have `start` set `complete=false` and add a `stop` callback that will update the ES entry to `complete=true` if so.  
It is better to have that info available in the ES so the status is available for search query.